From: "Kirill A. Shutemov" <kirill@shutemov.name>
Subject: mm-treewide-redefine-max_order-sanely-fix-fix
Date: Thu, 6 Apr 2023 10:25:29 +0300

On Wed, Apr 05, 2023 at 10:20:26PM -0700, Guenter Roeck wrote:
>
> This patch results in various boot failures (hang) on arm targets
> in linux-next. Debug messages reveal the reason.
>
> ########### MAX_ORDER=10 start=0 __ffs(start)=-1 min()=10 min_t=-1
>                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>
> If start==0, __ffs(start) returns 0xfffffff or (as int) -1, which min_t()
> interprets as such, while min() apparently uses the returned unsigned long
> value. Obviously a negative order isn't received well by the rest of the
> code.

Actually, __ffs() is not defined for 0.

Maybe something like this?

Link: https://lkml.kernel.org/r/20230406072529.vupqyrzqnhyozeyh@box.shutemov.name
Signed-off-by: "Kirill A. Shutemov" <kirill@shutemov.name>
Cc: Mike Rapoport <rppt@kernel.org>
Reported-by: Guenter Roeck <linux@roeck-us.net>
  Link: https://lkml.kernel.org/r/9460377a-38aa-4f39-ad57-fb73725f92db@roeck-us.net
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/memblock.c       |    6 +++++-
 mm/memory_hotplug.c |    8 +++++++-
 2 files changed, 12 insertions(+), 2 deletions(-)

--- a/mm/memblock.c~mm-treewide-redefine-max_order-sanely-fix-fix
+++ a/mm/memblock.c
@@ -2043,7 +2043,11 @@ static void __init __free_pages_memory(u
 	int order;
 
 	while (start < end) {
-		order = min_t(int, MAX_ORDER, __ffs(start));
+		/* __ffs() behaviour is undefined for 0 */
+		if (start)
+			order = min_t(int, MAX_ORDER, __ffs(start));
+		else
+			order = MAX_ORDER;
 
 		while (start + (1UL << order) > end)
 			order--;
--- a/mm/memory_hotplug.c~mm-treewide-redefine-max_order-sanely-fix-fix
+++ a/mm/memory_hotplug.c
@@ -605,7 +605,13 @@ static void online_pages_range(unsigned
 	 * this and the first chunk to online will be pageblock_nr_pages.
 	 */
 	for (pfn = start_pfn; pfn < end_pfn;) {
-		int order = min_t(int, MAX_ORDER, __ffs(pfn));
+		int order;
+
+		/* __ffs() behaviour is undefined for 0 */
+		if (pfn)
+			order = min_t(int, MAX_ORDER, __ffs(pfn));
+		else
+			order = MAX_ORDER;
 
 		(*online_page_callback)(pfn_to_page(pfn), order);
 		pfn += (1UL << order);
_
